src/hotspot/share/jfr/leakprofiler/chains/dfsClosure.cpp

```diff
@@ -60,52 +63,91 @@ void DFSClosure::find_leaks_from_root_set(EdgeStore* edge_store,
   rs.process();
+  dfs.drain_probe_stack();
 }
 
 DFSClosure::DFSClosure(EdgeStore* edge_store, JFRBitSet* mark_bits, const Edge* start_edge)
   :_edge_store(edge_store), _mark_bits(mark_bits), _start_edge(start_edge),
-  _max_depth(max_dfs_depth), _depth(0), _ignore_root_set(false) {
+  _max_depth(max_dfs_depth), _depth(0), _ignore_root_set(false),
+  _probe_stack(MIN2(256UL, max_dfs_depth)) {
 }
 
-void DFSClosure::closure_impl(UnifiedOopRef reference, const oop pointee) {
-  assert(pointee != nullptr, "invariant");
-  assert(!reference.is_null(), "invariant");
-
-  if (GranularTimer::is_finished()) {
-    return;
-  }
-
-  if (_depth == 0 && _ignore_root_set) {
-    // Root set is already marked, but we want
-    // to continue, so skip is_marked check.
-    assert(_mark_bits->is_marked(pointee), "invariant");
-    _reference_stack[_depth] = reference;
-  } else {
-    if (_mark_bits->is_marked(pointee)) {
-      return;
+#ifdef ASSERT
+DFSClosure::~DFSClosure() {
+  assert(_probe_stack.is_empty() || GranularTimer::is_finished(),
+         "Should have drained the probe stack");
+}
+#endif // ASSERT
+
+void DFSClosure::drain_probe_stack() {
+
+  while (!_probe_stack.is_empty() &&
+         !GranularTimer::is_finished()) {
+
+    const ProbeStackItem psi = _probe_stack.pop();
+
+    const UnifiedOopRef reference = psi.r;
+    assert(!reference.is_null(), "invariant");
+    const oop pointee = reference.dereference();
+    assert(pointee != nullptr, "invariant");
+
+    _depth = psi.depth;
+
+    if (_depth == 0 && _ignore_root_set) {
+      // Root set is already marked, but we want
+      // to continue, so skip is_marked check.
+      assert(_mark_bits->is_marked(pointee), "invariant");
+      _reference_stack[_depth] = reference;
+    } else {
+      if (_mark_bits->is_marked(pointee)) {
+        continue;
+      }
+      _mark_bits->mark_obj(pointee);
+      _reference_stack[_depth] = reference;
+      // is the pointee a sample object?
+      if (pointee->mark().is_marked()) {
+        add_chain();
+      }
     }
-    _mark_bits->mark_obj(pointee);
-    _reference_stack[_depth] = reference;
-    // is the pointee a sample object?
-    if (pointee->mark().is_marked()) {
-      add_chain();
+    assert(_max_depth >= 1, "invariant");
+    if (_depth < _max_depth - 1) {
+      _depth++; // increase range for do_oop() to pick up
+
+      if (pointee->is_objArray()) {
+        objArrayOop pointee_oa = (objArrayOop)pointee;
+        const int len = pointee_oa->length();
+        // since our stack items are larger than those of GC marking stacks,
+        // we use a smaller stride
+        const int stridelen = MAX2((uintx)1, ObjArrayMarkingStride / 2);
+        const int begidx = psi.chunk * stridelen;
+        const int endidx = MIN2(len, (psi.chunk + 1) * stridelen);
+        if (endidx > begidx) {
+          if (endidx < len) {
+            ProbeStackItem psi2 = psi;
+            psi2.chunk ++;
+            _probe_stack.push(psi2);
```

Comment:
Could it be a problem that `pointee` has already been marked? To accomplish the striding, the same `pointee` needs to be revisited with the new chunk count to evaluate the next range.  However, the next time it's popped off the stack, it will get skipped over on line 108 since it's already been marked. 

I ran `TestJcmdDumpPathToGCRootsBFSDFS.java` and the test passes even after I add `assert(psi.chunk==0 )` in this block, which might indicate only the first range of each array is ever getting evaluated.